File: app/src/main/java/com/example/bluetoothapp/transformations/SleepStageClassifier.java
package com.example.bluetoothapp.transformations;

import android.os.Build;

import androidx.annotation.RequiresApi;

import java.util.ArrayList;

import classifier.FFT;

public class SleepStageClassifier {

    /*
    TODO: types to Objects ?
    TODO: double to integer ?
     */
    private double samplingFreq;
    private int inputLength;
    private int numSecondsInInput;
    private int fftLength;
    private FFT fftObject;
    private double[] basePSD;//values of PSD for awake state
    private double minDensity;//minimum threshold for density to be considered a peak


    @RequiresApi(api = Build.VERSION_CODES.N)
    public static void main(String[] args) throws Exception {
        /*
        !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
        Assume input is 4 seconds of data taken at sampling Rate = 1024 Hz
        so inputLength = 4 * 1024
        fftLength is 1024, since signal is periodic
        fs = 1024 (let it be same as original sampling Rate)
         */

        int inputLength = 4 * 1024;
        int fs = 1024;

        SleepStageClassifier clf = new SleepStageClassifier(fs, inputLength, 4);

        double[] calibrationSignal = new double[inputLength * 2];// 2 * 4 = 8 seconds of calibration
        for (int i = 0; i < inputLength * 2 ; i++) {
            // 13 Hz is frequency of sin wave, 2Hz is frequency of cos wave,
            calibrationSignal[i] = Math.sin(2 * Math.PI * 13 * i/fs) + Math.cos(2 * Math.PI * 2 * i/fs);
        }

        //here user waits for 8 seconds to calibrate software
        clf.calibrateBaseState(calibrationSignal, 8);

        double[] signal = new double[inputLength];//just inputLength for all other signals
        for (int i = 0; i < inputLength; i++) {
            // higher amplitude at 13 Hz and a 5Hz, frequency 20 freq sin is redundant -> nonREM
            signal[i] = 2 *Math.sin(2 * Math.PI * 13 * i/fs) + Math.cos(2 * Math.PI * 20 * i/fs) + Math.cos(2 * Math.PI * 5 * i/fs);
        }
        String state = clf.getSleepStage(signal);
        System.out.println(state);

    }

    public SleepStageClassifier(double samplingFreqFromSource, int inputLength, int numSecondsInInput ){
        this.samplingFreq = samplingFreqFromSource;
        this.inputLength = inputLength;
        this.numSecondsInInput = numSecondsInInput;
        this.fftLength = inputLength / numSecondsInInput;
        this.fftObject = new FFT(inputLength, fftLength, samplingFreq);
    }


    public void calibrateBaseState(double[] values, int seconds) throws Exception {
        /*
        set base PSD values for brain in some base state(awake)
        All comparisons will be made with this state
        @param values is input array of signal with properties given in constructor
        BUT with length = inputLength * (seconds / numSecondsInInput)
        @param seconds is num of seconds to calibrate base state (about 10 seconds of calm awake state)
        must be >= numSecondsInInput and be its multiply
        */
        if(seconds % numSecondsInInput > 0 || seconds < numSecondsInInput){
            throw new Exception("seconds must be >= numSecondsInInput and be its multiply");
        }

        if(values.length != inputLength * seconds / numSecondsInInput){
            throw new Exception("values.length != inputLength * (seconds / numSecondsInInput)");
        }

        ArrayList<double[]> basePSDs = new ArrayList<>();

        int numPSDs = values.length/inputLength;
        for(int split = 0; split < numPSDs; split++){
            double[] valuesPart = new double[inputLength];
            if (inputLength >= 0)
                System.arraycopy(values, split * inputLength, valuesPart, 0, inputLength);
            basePSDs.add(fftObject.computePSD(valuesPart));
        }

        int psdLen = basePSDs.get(0).length;
        basePSD = new double[psdLen];

        for(int i = 0; i < psdLen; i++){
            double curPSD = 0;
            for(double[] psd : basePSDs){
                curPSD += psd[i];
            }
            //basePSD[i] is mean of measured psds for ith frequency
            basePSD[i] = curPSD / basePSDs.size();
        }

        minDensity = 0;
        for(double ampl : basePSD){
            if (ampl > minDensity){
                minDensity = ampl;
            }
        }

    }

    /*
    We will possibly need to change input type to int[]
     */
    @RequiresApi(api = Build.VERSION_CODES.N)
    public String getSleepStage(double[] values) throws Exception {

        if (basePSD == null){
            throw new Exception("Perform calibration by calling calibrateBaseState");
        }
        double[] psd = fftObject.computePSD(values);

        ArrayList<Integer> freqsWithHighDensities = new ArrayList<Integer>();
        for(int i = 0; i < psd.length; i++){
            if (psd[i] > basePSD[i] && psd[i] > minDensity){
                System.out.println("Freq " + i);
                System.out.println("BasePSD: " + basePSD[i]);
                System.out.println("GotPSD: " + psd[i]);
                System.out.println("minDensity: " +minDensity);
                freqsWithHighDensities.add(i);
            }


        }

        return classifySleepStage(freqsWithHighDensities);


    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public String classifySleepStage(ArrayList<Integer> arrayOfFreqsWithPeaks){
        // array [1, 3, 10] corresponds to peaks at 1Hz, 3Hz, 10Hz
        //array is given sorted in ascending order
        arrayOfFreqsWithPeaks.removeIf(x -> (x >= 17));//freqs >= 17 are useless now

        String[] states = {"awake", "REM", "nonREM"};
        System.out.println(arrayOfFreqsWithPeaks);
        if(arrayOfFreqsWithPeaks.size() == 0){
            return states[0];
        }

        String state = "awake";
        for(int maxf : arrayOfFreqsWithPeaks){
            if(0 < maxf && maxf < 7){ // indicator of REM - sleep
                state = states[1];
            }
            if(12 < maxf && maxf < 17){
                state = states[2];
            }
        }

        return state;

    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public String classifyEyesStage(ArrayList<Integer> arrayOfFreqsWithPeaks){
        // array [1, 3, 10] corresponds to peaks at 1Hz, 3Hz, 10Hz
        //array is given sorted in ascending order
        arrayOfFreqsWithPeaks.removeIf(x -> (x >= 17));//freqs >= 17 are useless now

        String[] states = {"open", "closed"};
        System.out.println(arrayOfFreqsWithPeaks);
        if(arrayOfFreqsWithPeaks.size() == 0){
            return states[0];
        }

        String state = "awake";
        for(int maxf : arrayOfFreqsWithPeaks){
            if(8 < maxf && maxf < 15){ // indicator closed eyes
                state = states[1];
                break;
            }

        }

        return state;

    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public String getEyesStage(double[] values) throws Exception {

        if (basePSD == null){
            throw new Exception("Perform calibration by calling calibrateBaseState");
        }
        double[] psd = fftObject.computePSD(values);

        ArrayList<Integer> freqsWithHighDensities = new ArrayList<Integer>();
        for(int i = 0; i < psd.length; i++){
            if (psd[i] > basePSD[i] && psd[i] > minDensity){
                System.out.println("Freq " + i);
                System.out.println("BasePSD: " + basePSD[i]);
                System.out.println("GotPSD: " + psd[i]);
                System.out.println("minDensity: " +minDensity);
                freqsWithHighDensities.add(i);
            }


        }

        return classifySleepStage(freqsWithHighDensities);


    }
}

File: app/src/main/java/com/example/bluetoothapp/transformations/SleepStageClassifier.kt
package com.example.bluetoothapp.transformations
import kotlinx.android.synthetic.main.fragment_eyes_state.*
import android.os.Build
import android.widget.TextView
import androidx.annotation.RequiresApi
import classifier.FFT
import com.example.bluetoothapp.R
import com.example.bluetoothapp.SampleApplication
import kotlinx.android.synthetic.main.fragment_eyes_state.view.*
import java.util.*
import kotlin.math.cos
import kotlin.math.sin

class SleepStageClassifier(/*
    TODO: types to Objects ?
    TODO: double to integer ?
     */
    private val samplingFreq: Double,
    private val inputLength: Int,
    private val numSecondsInInput: Int
) {
    private val fftLength: Int = inputLength / numSecondsInInput
    private val fftObject: FFT
    private var basePSD //values of PSD for awake state
            : DoubleArray? = null
    private var minDensity //minimum threshold for density to be considered a peak
            = 0.0

    @Throws(Exception::class)
    fun calibrateBaseState(values: DoubleArray, seconds: Int) {
        /*
        set base PSD values for brain in some base state(awake)
        All comparisons will be made with this state
        @param values is input array of signal with properties given in constructor
        BUT with length = inputLength * (seconds / numSecondsInInput)
        @param seconds is num of seconds to calibrate base state (about 10 seconds of calm awake state)
        must be >= numSecondsInInput and be its multiply
        */
        if (seconds % numSecondsInInput > 0 || seconds < numSecondsInInput) {
            throw Exception("seconds must be >= numSecondsInInput and be its multiply")
        }
        if (values.size != inputLength * seconds / numSecondsInInput) {
            throw Exception("values.length != inputLength * (seconds / numSecondsInInput)")
        }
        val basePSDs = ArrayList<DoubleArray>()
        val numPSDs = values.size / inputLength
        for (split in 0 until numPSDs) {
            val valuesPart = DoubleArray(inputLength)
            if (inputLength >= 0) System.arraycopy(
                values,
                split * inputLength,
                valuesPart,
                0,
                inputLength
            )
            basePSDs.add(fftObject.computePSD(valuesPart))
        }
        val psdLen: Int = basePSDs[0].size
        basePSD = DoubleArray(psdLen)
        for (i in 0 until psdLen) {
            var curPSD = 0.0
            for (psd in basePSDs) {
                curPSD += psd[i]
            }
            //basePSD[i] is mean of measured psds for ith frequency
            basePSD!![i] = curPSD / basePSDs.size
        }
        minDensity = 0.0
        for (ampl in basePSD!!) {
            if (ampl > minDensity) {
                minDensity = ampl
            }
        }
    }

    /*
    We will possibly need to change input type to int[]
     */
    @RequiresApi(api = Build.VERSION_CODES.N)
    @Throws(Exception::class)
    fun getSleepStage(values: DoubleArray?): String {
        if (basePSD == null) {
            throw Exception("Perform calibration by calling calibrateBaseState")
        }
        val psd = fftObject.computePSD(values!!)
        val freqsWithHighDensities = ArrayList<Int>()
        for (i in psd.indices) {
            if (psd[i] > basePSD!![i] && psd[i] > minDensity) {
                println("Freq $i")
                println("BasePSD: " + basePSD!![i])
                println("GotPSD: " + psd[i])
                println("minDensity: $minDensity")
                freqsWithHighDensities.add(i)
            }
        }
        return classifySleepStage(freqsWithHighDensities)
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    fun classifySleepStage(arrayOfFreqsWithPeaks: ArrayList<Int>): String {
        // array [1, 3, 10] corresponds to peaks at 1Hz, 3Hz, 10Hz
        //array is given sorted in ascending order
        arrayOfFreqsWithPeaks.removeIf { x: Int -> x >= 17 } //freqs >= 17 are useless now
        val states = arrayOf("awake", "REM", "nonREM")
        println(arrayOfFreqsWithPeaks)
        if (arrayOfFreqsWithPeaks.size == 0) {
            return states[0]
        }
        var state = "awake"
        for (maxf in arrayOfFreqsWithPeaks) {
            if (maxf in 1..6) { // indicator of REM - sleep
                state = states[1]
            }
            if (maxf in 13..16) {
                state = states[2]
            }
        }
        return state
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    fun classifyEyesStage(arrayOfFreqsWithPeaks: ArrayList<Int>): String {
        // array [1, 3, 10] corresponds to peaks at 1Hz, 3Hz, 10Hz
        //array is given sorted in ascending order
        arrayOfFreqsWithPeaks.removeIf { x: Int -> x >= 17 } //freqs >= 17 are useless now
        val states = arrayOf("open", "closed")
        println(arrayOfFreqsWithPeaks)
        if (arrayOfFreqsWithPeaks.size == 0) {
            return states[0]
        }
        var state = "awake"
        for (maxf in arrayOfFreqsWithPeaks) {
            if (maxf in 9..14) { // indicator closed eyes
                state = states[1]
                break
            }
        }
        SampleApplication.eyesOpen = state
        return state
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @Throws(Exception::class)
    fun getEyesStage(values: DoubleArray?): String {
        if (basePSD == null) {
            throw Exception("Perform calibration by calling calibrateBaseState")
        }
        val psd = fftObject.computePSD(values!!)
        val freqsWithHighDensities = ArrayList<Int>()
        for (i in psd.indices) {
            if (psd[i] > basePSD!![i] && psd[i] > minDensity) {
                println("Freq $i")
                println("BasePSD: " + basePSD!![i])
                println("GotPSD: " + psd[i])
                println("minDensity: $minDensity")
                freqsWithHighDensities.add(i)
            }
        }
        return classifySleepStage(freqsWithHighDensities)
    }

    companion object {
        @RequiresApi(api = Build.VERSION_CODES.N)
        @Throws(Exception::class)
        @JvmStatic
        fun main(args: Array<String>) {
            /*
        !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
        Assume input is 4 seconds of data taken at sampling Rate = 1024 Hz
        so inputLength = 4 * 1024
        fftLength is 1024, since signal is periodic
        fs = 1024 (let it be same as original sampling Rate)
         */
            val inputLength = 4 * 1024
            val fs = 1024
            val clf = SleepStageClassifier(fs.toDouble(), inputLength, 4)
            val calibrationSignal = DoubleArray(inputLength * 2) // 2 * 4 = 8 seconds of calibration
            for (i in 0 until inputLength * 2) {
                // 13 Hz is frequency of sin wave, 2Hz is frequency of cos wave,
                calibrationSignal[i] =
                    sin(2 * Math.PI * 13 * i / fs) + cos(2 * Math.PI * 2 * i / fs)
            }

            //here user waits for 8 seconds to calibrate software
            clf.calibrateBaseState(calibrationSignal, 8)
            val signal = DoubleArray(inputLength) //just inputLength for all other signals
            for (i in 0 until inputLength) {
                // higher amplitude at 13 Hz and a 5Hz, frequency 20 freq sin is redundant -> nonREM
                signal[i] =
                    2 * sin(2 * Math.PI * 13 * i / fs) + cos(2 * Math.PI * 20 * i / fs) + cos(
                        2 * Math.PI * 5 * i / fs
                    )
            }
            val state = clf.getSleepStage(signal)
            println(state)
        }
    }

    init {
        fftObject = FFT(inputLength, fftLength, samplingFreq)
    }
}